core/src/toga/__init__.py

```diff
@@ -89,12 +89,26 @@
     "WebView",
 ]
 
-# Examples of valid version strings
-# __version__ = '1.2.3.dev1'  # Development release 1
-# __version__ = '1.2.3a1'     # Alpha Release 1
-# __version__ = '1.2.3b1'     # Beta Release 1
-# __version__ = '1.2.3rc1'    # RC Release 1
-# __version__ = '1.2.3'       # Final Release
-# __version__ = '1.2.3.post1' # Post Release 1
 
-__version__ = "0.3.0.dev39"
+def package_version(file, name):
+    try:
+        # Read version from SCM metadata
+        # This will only exist in a development environment
+        from setuptools_scm import get_version
+
+        # Excluded from coverage because a pure test environment (such as the one
+        # used by tox in CI) won't have setuptools_scm
+        return get_version(root="../../..", relative_to=file)  # pragma: no cover
+    except (ModuleNotFoundError, LookupError):
+        # If setuptools_scm isn't in the environment, the call to import will fail.
+        # If it *is* in the environment, but the code isn't a git checkout (e.g.,
+        # it's been pip installed non-editable) the call to get_version() will fail.
+        # If either of these occurs, read version from the installer metadata.
+        from importlib.metadata import version
+
+        # The Toga package names as defined in setup.cfg all use dashes.
+        package = "toga-core" if name == "toga" else name.replace("_", "-")
+        return version(package)
```

Comment:
Given that we've just opened the door for end-user defined backends, would it be preferable to just pass in "toga_core" or "toga_cocoa" as `name`, rather than do this dynamic handling in code?

Reply:
This couldn't be used by third-party backends as it is, because would have no way of knowing what to pass for `root`. Since I didn't intend for this to be part of Toga's public API, I'll change its name to add a leading underscore.